### PR TITLE
Minimal kernel

### PR DESCRIPTION
**Pull request by MylesBorins**
Implements everything as documented in https://github.com/nodejs/modules/pull/180

Alternative to https://github.com/nodejs/ecmascript-modules/pull/2 and https://github.com/nodejs/ecmascript-modules/pull/3

### CODE REVIEW

File: doc/api/esm.md [hunk 1]
```diff
@@ -54,6 +54,18 @@ property:
...
+
+ESM must have the `.mjs` extension
+
+### Mandatory file extensions
+
+You must provide a file extension when using the `import` keyword.
+
+### No importing directories
+
+There is no support for importing directories.
```
Comment (ljharb) [inline doc/api/esm.md]: can we add something like "at this time" to these three, so there's no implication that this state is set in stone?

File: doc/api/esm.md [hunk 2]
```diff
@@ -145,7 +156,8 @@ CommonJS modules loaded.
 ### Resolve hook
 
 The resolve hook returns the resolved file URL and module format for a
-given module specifier and parent file URL:
+given module specifier
+ile URL:
```
Comment (ljharb) [inline doc/api/esm.md]: typo?

File: doc/api/esm.md [hunk 3]
```diff
@@ -78,31 +90,30 @@ Modules will be loaded multiple times if the `import` specifier used to resolve
...
 
 ```js
 // foo.js
 module.exports = { one: 1 };
 
-// bar.js
-import foo from './foo.js';
+// bar.mjs
+import { createRequireFromPath } from 'module';
+const require = createRequireFromPath(import.meta.url.replace('file://', ''));
```
Comment (MylesBorins) [inline doc/api/esm.md]: THERE HAS TO BE A BETTER WAY

File: test/common/index.mjs [hunk 4]
```diff
@@ -1,6 +1,14 @@
 // Flags: --experimental-modules
 /* eslint-disable node-core/required-modules */
-import common from './index.js';
+
+import { createRequireFromPath } from 'module';
+
+function createRequire(url) {
+  return createRequireFromPath(url.replace('file://', ''));
```
Comment (MylesBorins) [inline test/common/index.mjs]: another place to fix this if it isn't the best way to do it... it works but makes me sad.

File: doc/api/esm.md (hunk 1 above)
Comment (MylesBorins) [reply doc/api/esm.md]: I've added documentation at the top of the file about the minimal kernel and expectations regarding feature sets. Does that satisfy this concern?

**Comment by MylesBorins** [comment]
CI: https://ci.nodejs.org/job/node-test-pull-request/17573/

File: doc/api/esm.md (hunk 3 above)
Comment (michael-ciniawsky) [reply doc/api/esm.md]:
```
createRequire(import.meta.url)
```

`import.meta.url` is **fully** qualified, so it's possible to handle it in a `createRequire()` `{Function}` instead, I personally think that use cases for URL's which aren't fully qualified are rare... but could be wrong.

Comment (targos) [reply doc/api/esm.md]: I thought we added something like `createRequireFromURL` :thinking:

File: doc/api/esm.md (hunk 1 above)
Comment (ljharb) [reply doc/api/esm.md]: I suppose it can; it's not a huge deal but it'd probably be nice to restate it.

File: doc/api/esm.md [hunk 5]
```diff
@@ -175,8 +175,6 @@ module. This can be one of the following:
 | --- | --- |
 | `'esm'` | Load a standard JavaScript module |
 | `'cjs'` | Load a node-style CommonJS module |
-| `'builtin'` | Load a node builtin CommonJS module |
```
Comment (targos) [inline doc/api/esm.md]: this line should stay

Comment (MylesBorins) [reply doc/api/esm.md]: the minimal kernel removes support for importing native modules

File: doc/api/esm.md (hunk 3 above)
Comment (michael-ciniawsky) [reply doc/api/esm.md]: I think it was `createRequireFromPath` and a potential `createRequirefromURL` was planned for the minimal kernel at first

File: doc/api/esm.md (hunk 1 above)
Comment (MylesBorins) [reply doc/api/esm.md]: If we repeat it here we have to for every feature that may change... would get redundant imho

File: doc/api/esm.md (hunk 3 above)
Comment (michael-ciniawsky) [reply doc/api/esm.md]: `createRequire` definitely would have been 'nicer' API wise and possible to be implemented as such

Comment (MylesBorins) [reply doc/api/esm.md]: we have not yet shipped `createRequireFromPath` so if we want to make any changes we should do so ASAP.

We can add `createRequireFromURL` as an iteration if we don't intend to change master right now. 

/cc @jasnell @devsnek

File: doc/api/esm.md [hunk 6]
```diff
@@ -54,6 +54,10 @@ property:
 
 ## Notable differences between `import` and `require`
 
+### Only Support for .mjs
+
+ESM must have the `.mjs` extension
```
Comment (targos) [inline doc/api/esm.md]: Missing period.

File: doc/api/esm.md [hunk 7]
```diff
@@ -145,7 +149,7 @@ CommonJS modules loaded.
 ### Resolve hook
 
 The resolve hook returns the resolved file URL and module format for a
-given module specifier and parent file URL:
+given module specifier and parent file URL
```
Comment (targos) [inline doc/api/esm.md]: If you remove the colon, I think there should be a period.

File: doc/api/esm.md (hunk 3 above)
Comment (michael-ciniawsky) [reply doc/api/esm.md]: I don't wintende to sound 'rude', but i'm under the impression that it's not _really_ needed to split `createRequire` into two methods for the majority of use cases. Which use cases do exists which require support for unqualified URLs for e.g? The (main) use cases I can think of atm are 

- `import.meta.url`
- `file://`

And for files it should be treated as either `file://to/file.ext` (File URL (fully qualified)) or `to/file.ext` (File Path) imho

**Comment by MylesBorins** [comment]
Test release can be found at https://nodejs.org/download/test/v11.0.0-test2018100224d19875fb/

Install via nvm:

```
$ NVM_NODEJS_ORG_MIRROR=https://nodejs.org/download/test nvm install v11.0.0-test2018100224d19875fb
```

It may fail right this moment as it is still building

File: doc/api/esm.md (hunk 3 above)
Comment (MylesBorins) [reply doc/api/esm.md]: We should move conversation as this one got folded

File: doc/api/esm.md [hunk 8]
```diff
@@ -78,31 +91,30 @@ Modules will be loaded multiple times if the `import` specifier used to resolve
...
 
 ```js
 // foo.js
 module.exports = { one: 1 };
 
-// bar.js
-import foo from './foo.js';
+// bar.mjs
+import { createRequireFromPath } from 'module';
+const require = createRequireFromPath(new URL(import.meta.url).pathname);
```
Comment (MylesBorins) [inline doc/api/esm.md]: continued conversation from https://github.com/nodejs/ecmascript-modules/pull/6#discussion_r221830385 about ergonomics of `createRequireFromPath`.

import.meta.url is a url string, starts with `file://`, don't like having to create this URL object just for this.

File: test/common/index.mjs [hunk 9]
```diff
@@ -2,9 +2,10 @@
 /* eslint-disable node-core/required-modules */
 
 import { createRequireFromPath } from 'module';
+import { fileURLToPath as toPath } from 'url';
 
-function createRequire(url) {
-  return createRequireFromPath(new URL(url).pathname);
+function createRequire(metaUrl) {
+  return createRequireFromPath(toPath(metaUrl));
```
Comment (MylesBorins) [inline test/common/index.mjs]: perhaps 3rd time is the charm.

now using `url.fileURLToPath` (thanks @guybedford)

This seems reasonable enough for the kernel but an obvious first place to improve the UX

File: doc/api/esm.md [hunk 10]
```diff
@@ -107,7 +107,8 @@ module.exports = { one: 1 };
 
 // bar.mjs
 import { createRequireFromPath } from 'module';
-const require = createRequireFromPath(new URL(import.meta.url).pathname);
+import { fileURLToPath as toPath } from 'url';
+const require = createRequireFromPath(toPath(import.meta.url));
```
Comment (MylesBorins) [inline doc/api/esm.md]: perhaps 3rd time is the charm.

now using `url.fileURLToPath` (thanks @guybedford)

This seems reasonable enough for the kernel but an obvious first place to improve the UX

**Comment by MylesBorins** [comment]
ci: https://ci.nodejs.org/job/node-test-pull-request/17581/

File: doc/api/esm.md [hunk 11]
```diff
@@ -78,31 +91,32 @@ Modules will be loaded multiple times if the `import` specifier used to resolve
...
-import './foo?query=1'; // loads ./foo with query of "?query=1"
-import './foo?query=2'; // loads ./foo with query of "?query=2"
+import './foo.mjs?query=1'; // loads ./foo.mjs with query of "?query=1"
+import './foo.mjs?query=2'; // loads ./foo.mjs with query of "?query=2"
 ```
 
 For now, only modules using the `file:` protocol can be loaded.
 
-## Interop with existing modules
+## CommonJS,  JSON, and Native Modules
```
Comment (jdalton) [inline doc/api/esm.md]: ☝️ you have an extra space between CommonJS and JSON.

File: lib/internal/modules/cjs/loader.js [hunk 12]
```diff
@@ -43,6 +43,7 @@ const {
 const preserveSymlinks = !!process.binding('config').preserveSymlinks;
 const preserveSymlinksMain = !!process.binding('config').preserveSymlinksMain;
 const experimentalModules = !!process.binding('config').experimentalModules;
+const hasLoader = !!process.binding('config').userLoader;
```
Comment (jdalton) [inline lib/internal/modules/cjs/loader.js]: ☝️ is `--loader` flag being included in the minimal implementation?

Comment (MylesBorins) [reply lib/internal/modules/cjs/loader.js]: The plan was to include it in the kernel. There are no competing loader implementations and almost every proposal included loaders afaict

Comment (jdalton) [reply lib/internal/modules/cjs/loader.js]: I see. That wasn't covered by https://github.com/nodejs/modules/pull/180. I think there may be folks that want to experiment in the loader space. I believe providing one up-front will guide folks down a specific route. I don't see the `--loader` option as mission critical for a minimal barebones implementation.

**Comment by MylesBorins** [comment]
We agreed to land this in a meeting with quorum today

One last CI: https://ci.nodejs.org/job/node-test-pull-request/17605/

Will land if green

**Comment by MylesBorins** [comment]
landed in cc31d8b2d4...c9dafb9b71

**Comment by devsnek** [comment]
wait... why does this remove `vm.SourceTextModule`? that has nothing to do with the rest of this.

**Comment by jdalton** [comment]
@devsnek See https://github.com/nodejs/modules/pull/180#pullrequestreview-161213066.

**Pull request merged**